### ISSUE TITLE
Bug in bounding box display?

### ISSUE THREAD

**Issue by saintaardvark** [alankbi/detecto#75]
**Describe the bug**
It looks like bounding boxes are not displayed correctly in all circumstances.  In my case, a sheared image generated by [Roboflow](https://roboflow.com), with adjusted bounding boxes in VOC XML format, appear incorrectly when displayed with Detecto; the same bounding boxes do appear correctly when displayed with imgaug.

**Code and Data**
I've created a notebook demonstrating the issue [here](https://github.com/saintaardvark/detecto_bug/blob/main/detecto_bounding_box_bug.ipynb); the repo has the source image and VOC XML file.

Detecto's version:
![image](https://user-images.githubusercontent.com/1486793/109341609-7fe58680-781f-11eb-9e45-cb28b1d39f0a.png)

The colours make it a bit hard to see what's going on, but:

* There's a mixing bowl centered around x=200, y=100, but the mixing bowl box apears to be off; the lower right-hand corner of the box and its label are at approximately x=150, y=250.

* There's a plate oriented horizontally at approximately x=200 to 300, y=200; but its bounding box is oriented vertically, with its lower right-hand corner at approximately x=200, y=325.

 * There are a couple bowls at approximately x=100, y=150 to 250. The bounding boxes are a little unclear to me, so I'll leave them out of this discussion.

Imgaug's version:

![image](https://user-images.githubusercontent.com/1486793/109341515-604e5e00-781f-11eb-9703-f04288ac14fc.png)

- The mixing bowl box is oriented correctly and centered on the bowl itself.
-  Same for the plate box.
- The bowl boxes are also correctly oriented and centered.

**Environment:**
 - OS: Linux, Fedora 33
 - Python version: 3.9.1
 - Detecto version: 1.2.0
 - torch version: 1.7.1
 - torchvision version: 0.8.2 

**Additional context**
The notebook and image files can be found at https://github.com/saintaardvark/detecto_bug/.  

Thanks for a great library, and please let me know if you need any further info.

**Comment by alankbi** [comment]
I just ran this locally roughly the same code and image/label file and it seems to work fine - could you triple-check that your Detecto version is the latest at 1.2.0? I think in the past there was a bug with how the xmin, ymin, etc. were read in, but it should all be fixed in the latest version. 

If it's the latest, then could you try printing the contents of `targets` and reporting what that returns for the box coordinates?

**Comment by saintaardvark** [comment]
I've double checked that the version is 1.2.0.  (This is pinned in my requirements.txt for the notebook, and I've recreated the virtualenv just to be sure.) 

Printing out `targets` in the notebook gives me this:

```
{'boxes': tensor([[ 57, 147, 158, 218],
         [ 67, 154, 124, 184],
         [145, 248,  67, 150],
         [193, 314, 178, 230]]),
 'labels': ['bowl', 'bowl', 'mixing bowl', 'plate']}
```

I've done a bit more digging, and I _think_ I see what the problem might be.  I'll focus on [the first object in the `bug_demo.xml` file](https://github.com/saintaardvark/detecto_bug/blob/main/dataset/bug_demo.xml#L14-L26):

```xml
	<object>
		<name>bowl</name>
		<pose>Unspecified</pose>
		<truncated>0</truncated>
		<difficult>0</difficult>
		<occluded>0</occluded>
		<bndbox>
			<xmin>57</xmin>
			<xmax>147</xmax>
			<ymin>158</ymin>
			<ymax>218</ymax>
		</bndbox>
	</object> 
```
For `targets` up above, you should be able to get that like so:

```python
print targets['boxes'][0]
tensor([ 57, 147, 158, 218])
```

Looking at the code for [`show_labelled_image()`](https://github.com/alankbi/detecto/blob/master/detecto/visualize.py#L297-L302), I see this:

```python
    for i in range(boxes.shape[0]):
        box = boxes[i]
        width, height = (box[2] - box[0]).item(), (box[3] - box[1]).item()
        initial_pos = (box[0].item(), box[1].item())
        rect = patches.Rectangle(initial_pos,  width, height, linewidth=1,
                                 edgecolor='r', facecolor='none')
```


The two lines in the middle are the crucial bit. They appear to assume that:

* box[2] == xmax, and box[0] == xmin
* box[3] == ymax, and box[1] == ymin

However, the targets variable shows that these coordinates are stored differently:

* box[2] == ymin (not xmax)
* box[1] == xmax (not ymin)

Another way to say this: the box coordinates are _stored_ as:

```
[xmin, xmax, ymin, ymax]
```

We can see this from the printout of the `targets` variable up above; that placement matches the XML file (as in, each of the four corners is correct).

However, the box coordinates are _used_ in `show_labelled_image()` as:

```
[xmin, ymin, xmax, ymax]
```

(I'll try to add this to the notebook, as that would definitely be clearer...however, I'm having unrelated problems in my Python environment right now :facepalm:, so that might take a bit.)

Thanks again for looking at this!

**Comment by alankbi** [comment]
Good catch, and thanks for providing such detail! Could you maybe also take a look at the function [xml_to_csv](https://github.com/alankbi/detecto/blob/master/detecto/utils.py#L222) and share what the outputted CSV file from that looks like? 

The Dataset object makes use of that function internally, and as of the latest version it's been updated to read in coordinates in the `xmin, ymin, xmax, ymax` order regardless of how it's stored in the XML file. However, looking at your output, it seems to still be reading it in as `xmin, xmax, ymin, ymax` (the order stored in the XML file), which is definitely odd.

**Comment by saintaardvark** [comment]
Hi there -- finally got a chance to do this.  The updated note book is at https://github.com/saintaardvark/detecto_bug/blob/main/detecto_bounding_box_bug.ipynb, but the quick version: this bug has been fixed in [this commit](https://github.com/alankbi/detecto/commit/2938bab333997fd0696f5ece1dc504e30b0414eb), but that commit came *after* v1.2.0, and hasn't been pushed to PyPi et al.  In other words, the latest release available for install via pip doesn't have this fix.

Assuming that's correct, cutting a new release should take care of this.

Thanks again for looking into this!

**Comment by alankbi** [comment]
Just released a new version [v1.2.1](https://github.com/alankbi/detecto/releases/tag/v1.2.1). Feel free to take a look and let me know if that solves the issue!

**Comment by saintaardvark** [comment]
Awesome, that does the trick.  Here's the output of `show_labelled_image()` from Detecto 1.2.1:
![detecto-1 2 1](https://user-images.githubusercontent.com/1486793/110373690-a1572700-8004-11eb-809c-c332512019e1.png)
